integration-tests/tests/utils/config.py

```diff
@@ -69,6 +81,131 @@ def __post_init__(self) -> None:
             )
             raise ValueError(err_msg)
 
+        # Set clp_config_file_path and validate it.
+        object.__setattr__(
+            self, "clp_config_file_path", self.clp_package_dir / "etc" / "clp-config.yml"
+        )
+        validate_file_exists(self.clp_config_file_path)
+
+        # Set hostname.
+        object.__setattr__(self, "hostname", socket.gethostname())
+
+
+@dataclass(frozen=True)
+class PackageRun:
+    """Metadata for the running CLP package."""
+
+    #:
+    package_config: PackageConfig
+
+    #:
+    mode: Literal["clp-text", "clp-json"]
+
+    #:
+    clp_log_dir: Path = field(init=False, repr=True)
+
+    clp_run_config_file_path: Path = field(init=False, repr=True)
+
+    #:
+    clp_instance_id: str = field(init=False, repr=True)
+
+    def __post_init__(self) -> None:
+        """Validates the values specified at init, and initialises attributes."""
+        # Set clp_log_dir after validating it.
+        clp_log_dir = self.package_config.clp_package_dir / "var" / "log"
+        validate_dir_exists(clp_log_dir)
+        object.__setattr__(self, "clp_log_dir", clp_log_dir)
+
+        # Set clp_run_config_file_path after validating it.
+        clp_run_config_file_path = (
+            self.clp_log_dir / self.package_config.hostname / ".clp-config.yml"
+        )
+        validate_file_exists(clp_run_config_file_path)
+        object.__setattr__(self, "clp_run_config_file_path", clp_run_config_file_path)
+
+        # Set clp_instance_id.
+        clp_instance_id_file_path = self.clp_log_dir / self.package_config.hostname / "instance-id"
+        validate_file_exists(clp_instance_id_file_path)
+        clp_instance_id = self._get_clp_instance_id(clp_instance_id_file_path)
+        object.__setattr__(self, "clp_instance_id", clp_instance_id)
+
+        # Validate mode.
+        self._assert_mode_matches_run_config()
+
+    def _assert_mode_matches_run_config(self) -> None:
+        """Validates that self.mode matches the package values in the run config file."""
+        config_dict = self._load_run_config(self.clp_run_config_file_path)
+        config_mode = self._extract_mode_from_config(config_dict, self.clp_run_config_file_path)
+
+        if config_mode != self.mode:
+            error_msg = (
+                f"Mode mismatch: the mode specified to the PackageRun object was {self.mode},"
+                f" but the package is running in {config_mode} mode."
+            )
+            raise ValueError(error_msg)
+
+    @staticmethod
+    def _load_run_config(path: Path) -> dict[str, Any]:
```

Comment:
this is useful. we could load a config template, and then modify the entries of our interest.
There's also a possibility that we may need to randomize the port numbers to avoid collisions.